### PR TITLE
Send box renderer

### LINKED ISSUE #3038: Add Backgroundcolor for SendBoxButton to defaultStyleOptions
Feature Request

My requirement is to set the BackgroundColor of the sendBoxButton.
Currently, the defaultStyleOptions.js does not support this and i had to fall back to using css selectors.
defaultStyleOptions.js seems to be the natural place for this.

Current solution:
`    #webchat > div > div > div.main > div > button{
        background-color: aqua;
    }`

Desired solution:
`
        const styleOptions = {
            sendBoxButtonsendBoxBackgroundColor: "aqua"
        };
`

### LINKED ISSUE #3526: Optimizing the documentation for the customization of the webchat.
## Feature Request

**Is your feature request related to a problem? Please describe.**
The feature request is based on [this question](https://github.com/microsoft/BotFramework-WebChat/issues/3490).

**Describe the suggestion or request in detail**
I find that there is some informations missing in [this documentation](https://github.com/microsoft/BotFramework-WebChat/tree/master/samples/06.recomposing-ui/d.plain-ui#customizing-web-chat-with-a-plain-html-ui) at point 3:

> Using business layers and some UI components: opting into our UI but replacing just a handful of components as needed.

To replace some of the components it is necessary to fork the repository. It would be nice if someone could add the following information to this sentence:

(Using business layers and some UI [...]) **(Its necessary to fork the repository and pull our updates into your custom - WebChat)**.

### PR DESCRIPTION
**Pull request by felixvanleeuwen**
> Fixes #3526 
> Fixes #3038 

## Description
Added a prop that is sent down into `BasicWebChat` and is used to render the send box if not undefined.

## Specific Changes
Added a prop to `BasicWebChat` `defaultProps`.
Added a check in the `BasicWebChat` render to use the new prop instead of rendering the `BasicSendBox` 

## Review Checklist

> This section is for contributors to review your work.

-  [ ] Accessibility reviewed (tab order, content readability, alt text, color contrast)
-  [ ] Browser and platform compatibilities reviewed
-  [ ] CSS styles reviewed (minimal rules, no `z-index`)
-  [ ] Documents reviewed (docs, samples, live demo)
-  [ ] Internationalization reviewed (strings, unit formatting)
-  [ ] `package.json` and `package-lock.json` reviewed
-  [ ] Security reviewed (no data URIs, check for nonce leak)
-  [ ] Tests reviewed (coverage, legitimacy)

### CODE REVIEW

**Comment by corinagum** [comment]
Thanks for the PR!! 

1. Please fill out the CLA
1. Could you add the passing snapshots to the committed files? They need to be present in order for the CI to pass.

I'll take a closer look at this sometime this week.

File: packages/component/src/BasicWebChat.js
Comment (corinagum) [inline packages/component/src/BasicWebChat.js:89]: Rather than renderer, this will need to be changed to `sendBoxMiddleware`.
image
https://github.com/microsoft/BotFramework-WebChat/tree/master/packages/component/src/Middleware

I added a new issue to get rid of our deprecations to reduce confusion: https://github.com/microsoft/BotFramework-WebChat/issues/3534

@compulim, do you have thoughts?

**Review by corinagum — CHANGES_REQUESTED** [review]
This will need to be updated to use the middleware format rather than the deprecated renderer format. Please address that and above requests as well. :)

@compulim feel free to add. Do you think this is a good middle step while we wait for the new sendbox design?

File: packages/component/src/BasicWebChat.js
Comment (felixvanleeuwen) [reply packages/component/src/BasicWebChat.js]: I've changed the name of the prop.

"Middleware" implies that can it can be inserted in between existing logic.
This prop is very much an alternative route that can be used, once used it can't pass on to existing logic.

Having said that, do you still feel this is the correct name for the prop?
Or did you mean my solution should be a middleware and I have to change my implementation?

Comment (corinagum) [reply packages/component/src/BasicWebChat.js]: The implementation should be middleware. :) I think there are other considerations that @compulim would like to comment on as well. Pinging him.

**Comment by felixvanleeuwen** [comment]
I've updated the snaps but the CI still fails.
When I try to check the "Details" I get a 401.

How do I check why my CI is failing?

**Comment by corinagum** [comment]
@felixvanleeuwen thanks for the updates! Unfortunately non-internal developers are unable to check the CI. Some of them are DL timeouts, so I'm rerunning right now and after that I'll make a list of the failing tests. Thanks for your patience.

**Comment by corinagum** [comment]
@compulim ping

**Comment by felixvanleeuwen** [comment]
Any news on the failing CI?

**Comment by corinagum** [comment]
Both failing tests were issues resolved yesterday. Merged main in and restarted CI.

**Comment by felixvanleeuwen** [comment]
CI still fails, anything I can solve?

**Comment by corinagum** [comment]
CI is failing on my PR too -- both have lots of failures and look to be the same. Something else is going on, which I'll investigate today. Are you able to run the tests locally?
Will you be able to change to middleware for this PR?

**Comment by felixvanleeuwen** [comment]
>Are you able to run the tests locally?
Yes I can. I have to run them on a machine that is different from my regular work station so it'll take some extra effort.
>Will you be able to change to middleware for this PR?
In the current way it's implemented returning `false` from the `sendBoxMiddleware` would have the `BasicWebChat` default to the `BasicSendBox`.
How would you like me to change it?

**Comment by compulim** [comment]
Will get back today, I got some ideas on how to make it better.

**Comment by felixvanleeuwen** [comment]
I think I can look at any `createCoreMiddleware` to get a general idea of how it should work.
Hopefully I can set some time aside next week to work on this.

**Comment by felixvanleeuwen** [comment]
Starting work on this

**Comment by felixvanleeuwen** [comment]
@compulim Is this how you imagined it?

**Comment by corinagum** [comment]
@felixvanleeuwen Thanks for the updates! FYI, compulim is currently on vacation and will be unavailable for a few weeks. Since he wanted to go through the code, I'll go through and possibly leave a few comments, but we'll wait until he's back for the approval.

We just finished a release milestone, and our next is scheduled for next year, so presumably we'll still be able to get this into the next release.

Thanks for your patience!

File: samples/02.branding-styling-and-customization/l.idiosyncratic-send-box-rendering/README.md
Comment (corinagum) [inline samples/02.branding-styling-and-customization/l.idiosyncratic-send-box-rendering/README.md:7]: Update URL

**Comment by corinagum** [comment]
You don't need to do this immediately, since there will be frequent changes to `CHANGELOG.md` that will cause merge conflicts, but before this gets merged in, you will need to give yourself credit and add the changes to the Changelog :) I would include a line under `Added` in `Unreleased` for the SendBox changes, and an entry under `Samples` for the new sample. 

Please also add this sample to the /samples/README.md table. Thanks again :)

File: packages/component/src/BasicWebChat.js
Comment (felixvanleeuwen) [reply packages/component/src/BasicWebChat.js]: @compulim is the change satisfactory?

**Comment by felixvanleeuwen** [comment]
Hello, is there anything blocking about this PR still?
I'd like to hear what's wrong so I can take steps to remedy.

**Comment by felixvanleeuwen** [comment]
Hello?

**Comment by corinagum** [comment]
Hi @felixvanleeuwen sorry I missed your reply. We still need @compulim to do a pass review, but there are still issues unaddresed above that should be fixed. Please take a look at those.

**Comment by felixvanleeuwen** [comment]
I think all of the issues have been addressed.
If you feel that some have not been addressed: could you specifically list them?

**Comment by corinagum** [comment]
Reopening - I'm assuming close was unintentional :) Please let me know if not.

**Comment by sibbl** [comment]
Sorry for bringing up this PR again, but is there still any interest in being able to implement a custom search box? From a developers perspective it makes perfect sense to have a middleware for this just like we have for all other crucial parts of the UI.

Or is there some sample of solving it with current tools, as I saw that there's a `hideSendBox` style option?

**Comment by compulim** [comment]
Obsoleted by #5120.

**Pull request closed without merging**
